library/std/src/path.rs

```diff
@@ -1976,6 +1984,12 @@ impl Path {
         unsafe { &*(s.as_ref() as *const OsStr as *const Path) }
     }
 
+    fn from_inner_mut(inner: &mut OsStr) -> &mut Path {
+        // SAFETY: Path is just a wrapper around OsStr,
+        // therefore converting &mut OsStr to &mut Path is safe.
+        unsafe { &mut *(inner as *mut OsStr as *mut Path) }
```

Comment:
I don't see a `#[repr(transparent)]` on the struct definition - I don't think this transmute is correct.

Reply:
This is correct for standard library code.

Reply:
@dtolnay Can you say more? What property of std lets it do this?

Reply:
The fact that if any future compiler change caused these types to be given different layout or disrupted the present behavior of this code in any other way, the same change would atomically adjust std to accommodate that. In other words the standard library is written to function under exactly one compiler version ever, whereas downstream code is required to be sound under all possible future compilers which are non-breaking evolutions of the current implicit "spec".